src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -26,114 +28,239 @@
+            String commandType = fields[COMMAND_TYPE_INDEX].toLowerCase();
 
-            String field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
+            String errorMessage = validateCommand(fields, command);
+            if(!errorMessage.isEmpty()){
+                return createHttpError(errorMessage, HttpStatus.BAD_REQUEST);
             }
-            i++;
 
-            pet.setName(field);
+            Date date = parseDate(dateString);
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
+            int ownerID = Integer.parseInt(ownerIDString);
+            Owner owner;
 
             try {
-                pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(field));
-            } catch (ParseException e) {
-                HttpHeaders headers = new HttpHeaders();
-                headers.add("errors", "date " + field + " not valid");
-                return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
+                owner = clinicService.findOwnerById(ownerID);
+            }catch (ObjectRetrievalFailureException|EmptyResultDataAccessException e){
+                errorMessage = "In command: " + command + "\n";
+                errorMessage += "Owner is not found, given this ID: " + ownerID;
+                return createHttpError(errorMessage, HttpStatus.NOT_FOUND);
             }
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
+            int typeID = Integer.parseInt(typeIDString);
+            PetType type;
+
+            try {
+                type = clinicService.findPetTypeById(typeID);
+            }catch (ObjectRetrievalFailureException|EmptyResultDataAccessException e){
+                errorMessage = "In command: " + command + "\n";
+                errorMessage += "Pet type is not found, given this ID: " + typeID;
+                return createHttpError(errorMessage, HttpStatus.NOT_FOUND);
             }
-            i++;
 
-            if (pet != null) {
-                ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
-                for (int j = 0; j < ts.size(); j++) {
-                    if (ts.get(j).getName().toLowerCase().equals(field)) {
-                        pet.setType(ts.get(j));
+            Pet pet = createPet(name, date, type);
+
+            // Beware of setting owner before deletion. This will result in no deletion occurring.
+            if(commandType.equals("add")){
+                pet.setOwner(owner);
+                clinicService.savePet(pet);
+            }else{ // already validated command types, and we only have two commands. This should be the delete command.
+                List<Pet> pets = owner.getPets();
```

Comment:
if the list of pets is very long this might have a bad performance - let the database do the job of finding the pet, there are optimized for such things

Reply:
I wanted to, but I think I need the pet id in order to do so, and the csv doesn't provide it, only the name and the owner.